FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/FhirToDataLakeProcessingJobSpliter.cs

```diff
@@ -0,0 +1,291 @@
+using Microsoft.Health.Fhir.Synapse.Core.Jobs.Models;
+using Microsoft.Health.Fhir.Synapse.DataClient;
+using Microsoft.Health.Fhir.Synapse.DataClient.Api.Fhir;
+using Microsoft.Health.Fhir.Synapse.DataClient.Extensions;
+using Microsoft.Health.Fhir.Synapse.DataClient.Models.FhirApiOption;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Health.Fhir.Synapse.Core.Jobs
+{
+    public class FhirToDataLakeProcessingJobSpliter
+    {
+        private readonly IApiDataClient _dataClient;
+        private readonly IDiagnosticLogger _diagnosticLogger;
+        private readonly ILogger<FhirToDataLakeProcessingJobSpliter> _logger;
+
+        public FhirToDataLakeProcessingJobSpliter(
+            IApiDataClient dataClient,
+            IDiagnosticLogger diagnosticLogger,
+            ILogger<FhirToDataLakeProcessingJobSpliter> logger)
+        {
+            _dataClient = EnsureArg.IsNotNull(dataClient, nameof(dataClient));
+            _diagnosticLogger = EnsureArg.IsNotNull(diagnosticLogger, nameof(diagnosticLogger));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+        }
+
+        public int LowBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.LowBoundOfProcessingJobResourceCount;
+
+        public int HighBoundOfProcessingJobResourceCount { get; set; } = JobConfigurationConstants.HighBoundOfProcessingJobResourceCount;
+
+        public async IAsyncEnumerable<SubJobInfo> SplitJobAsync(string resourceType, DateTimeOffset? startTime, DateTimeOffset endTime, [EnumeratorCancellation] CancellationToken cancellationToken)
+        {
+            var totalCount = await GetResourceCountAsync(resourceType, startTime, endTime, cancellationToken);
+
+            if (totalCount < HighBoundOfProcessingJobResourceCount)
+            {
+                // Return job with total count lower than high bound.
+                _logger.LogInformation($"Generate one {resourceType} job with {totalCount} count.");
+
+                yield return new SubJobInfo
+                {
+                    ResourceType = resourceType,
+                    TimeRange = new TimeRange() { DataStartTime = startTime, DataEndTime = endTime },
+                    JobSize = totalCount,
+                };
+            }
+            else
+            {
+                // Split large size job using binary search.
+                var splitingStartTime = DateTimeOffset.Now;
+                _logger.LogInformation($"Start spliting {resourceType} job, total {totalCount} resources.");
+
+                var anchorList = await InitializeAnchorListAsync(resourceType, startTime, endTime, totalCount, cancellationToken);
+
+                _logger.LogInformation($"Spliting {resourceType} job. Use {(DateTimeOffset.Now - splitingStartTime).TotalMilliseconds} milliseconds to initilize anchor list.");
+                var lastSplitTimestamp = DateTimeOffset.Now;
+                DateTimeOffset? nextJobEnd = null;
```

Comment:
rename to lastjobend?

Reply:
Already has a lastEndTime variable to store last end time stamp here.